# Основное задание: Изучение библиотечных функций

### Требования: keras.optimizers, torch.optim: SGD, и модификации SGD (Nesterov, Momentum, AdaGrad, RMSProp, Adam). Изучите параметры вызываемых библиотечных функций.

# Изучение библиотеки _torch_

##### Выбираем как _target_ alcohol и обучаем модель на датасете

### ***SGD***:

In [1]:
from libs.chooseLib import choose_model
import numpy as np

if __name__ == '__main__':
    sample_raw = np.asarray(
            [[7.4, 0.70, 0.00, 1.9, 0.076, 11, 34, 0.9978, 3.51, 0.56, 5]]
        )

    model, scaler = choose_model('alcohol',"torch", "SGD")
    sample_scaled = scaler.transform(sample_raw)
    print(model(sample_scaled))

2025-05-30 16:15:36.891361: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1748610937.016600    6677 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1748610937.051117    6677 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1748610937.313403    6677 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1748610937.313428    6677 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1748610937.313431    6677 computation_placer.cc:177] computation placer alr

[DataLoader] shape=(1599, 12)
MSE: 0.35177, MAE: 0.45827, R²: 0.71744
9.61831


### ***Momentum***:

In [2]:
    model, _ = choose_model('alcohol',"torch", "Momentum")
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)
MSE: 0.34634, MAE: 0.45564, R²: 0.72180
9.653122


### ***Nesterov***:

In [3]:
    model, _ = choose_model('alcohol',"torch", "Nesterov")
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)
MSE: 0.34511, MAE: 0.45150, R²: 0.72278
9.624101


### ***RMSProp***:

In [4]:
    model, _ = choose_model('alcohol',"torch", "RMSProp")
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)
MSE: 0.36759, MAE: 0.45572, R²: 0.70473
9.388223


### ***Adam***:

In [5]:
    model, _ = choose_model('alcohol',"torch", "Adam", lr=0.5)
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)
MSE: 0.34983, MAE: 0.46071, R²: 0.71899
9.679335


### ***AdaGrad***:

In [6]:
    model, _ = choose_model('alcohol',"torch", "AdaGrad", lr=0.9999991)
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)
MSE: 0.43611, MAE: 0.49340, R²: 0.64969
9.313658


---
_Real Value: 9.4_

| Optimizer    |     MSE      |     MAE      |     R²      | Predict Value (s) |
|--------------|--------------|--------------|-------------|-------------------|
| SGD          |   0.35582    |   0.46175    |   0.71418   | 9.615764          |
| Momentum     |   0.34623    |   0.45548    |   0.72189   | 9.652524          |
| Nesterov     |   0.34510    |   0.45149    |   0.72279   | 9.624092          |
| RMSProp      |   0.36799    |   0.45324    |   0.70441   | 9.463510          |
| Adam         |   0.34945    |   0.46032    |   0.71930   | 9.677068          |
| AdaGrad      |   0.47579    |   0.51806    |   0.61781   | 9.254324          |

---

# Изучаем библиотеку _keras_

### ***SGD***:

In [7]:
    model, _ = choose_model('alcohol',"tensorflow", "SGD")
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)


I0000 00:00:1748610944.407361    6677 gpu_device.cc:2019] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 6283 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 3070 Laptop GPU, pci bus id: 0000:01:00.0, compute capability: 8.6
/home/brx/PyCharmMiscProject/.venv/lib/python3.12/site-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
2025-05-30 16:15:44.779586: W external/local_xla/xla/service/gpu/llvm_gpu_backend/default/nvptx_libdevice_path.cc:40] Can't find libdevice directory ${CUDA_DIR}/nvvm/libdevice. This may result in compilation or runtime failures, if the program we try to run uses routines from libdevice.
Searched for CUDA in the following directories:
  ./cuda_sdk_lib
  ipykernel_launcher.runfiles/cuda_nvcc
  ipyk

InternalError: {{function_node __wrapped__FloorMod_device_/job:localhost/replica:0/task:0/device:GPU:0}} 'cuLaunchKernel(function, gridX, gridY, gridZ, blockX, blockY, blockZ, 0, reinterpret_cast<CUstream>(stream), params, nullptr)' failed with 'CUDA_ERROR_INVALID_HANDLE' [Op:FloorMod] name: 

### ***Momentum***:

In [26]:
    model, _ = choose_model('alcohol',"tensorflow", "Momentum")
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)


/home/letooow/PycharmProjects/PythonProject/.venv/lib/python3.12/site-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 0, Loss: 23.6407337
Momentum: MSE = 0.6211, MAE = 0.5900, R² = 0.5011
9.90852


### ***Nesterov***:

In [27]:
    model, _ = choose_model('alcohol',"tensorflow", "Nesterov")
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)


/home/letooow/PycharmProjects/PythonProject/.venv/lib/python3.12/site-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 0, Loss: 0.54085505
Nesterov: MSE = 0.4023, MAE = 0.4855, R² = 0.6769
9.602156


### ***AdaGrad***:

In [31]:
    model, _ = choose_model('alcohol',"tensorflow", "AdaGrad", lr=0.9999991)
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)


/home/letooow/PycharmProjects/PythonProject/.venv/lib/python3.12/site-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 0, Loss: 18.9184189
AdaGrad: MSE = 0.3489, MAE = 0.4539, R² = 0.7198
9.646044


### ***RMSProp***:

In [29]:
    model, _ = choose_model('alcohol',"tensorflow", "RMSProp")
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)


/home/letooow/PycharmProjects/PythonProject/.venv/lib/python3.12/site-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 0, Loss: 62.4272346
RMSProp: MSE = 0.3856, MAE = 0.4913, R² = 0.6903
9.903184


### ***Adam***:

In [30]:
    model, _ = choose_model('alcohol',"tensorflow", "Adam")
    print(model(sample_scaled))

[DataLoader] shape=(1599, 12)


/home/letooow/PycharmProjects/PythonProject/.venv/lib/python3.12/site-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 0, Loss: 71.9465103
Adam: MSE = 0.3992, MAE = 0.4837, R² = 0.6793
9.348383


### Результат:

| Optimizer |   MSE   |   MAE   |   R²   | Predict Value |
|-----------|---------|---------|--------|---------------|
| SGD       | 0.3468  | 0.4530  | 0.7214 | 9.589417      |
| Momentum  | 0.6211  | 0.5900  | 0.5011 | 9.908520      |
| Nesterov  | 0.4023  | 0.4855  | 0.6769 | 9.602156      |
| AdaGrad   | 0.3489  | 0.4539  | 0.7198 | 9.646044      |
| RMSProp   | 0.3856  | 0.4913  | 0.6903 | 9.903184      |
| Adam      | 0.3992  | 0.4837  | 0.6793 | 9.348383      |



# Вывод



| Optimizer |  MSE (TF) | MSE (Torch) | Better MSE | R² (TF) | R² (Torch) | Better R² | MAE (TF) | MAE (Torch) | Better MAE | TF Predict | Torch Predict |
|-----------|-----------|-------------|------------|---------|------------|-----------|----------|-------------|------------|-------------|----------------|
| SGD       | 0.3468    | 0.35582     | TF         | 0.7214  | 0.71418    | TF        | 0.4530   | 0.46175     | TF         | 9.589417    | 9.615764       |
| Momentum  | 0.6211    | 0.34623     | Torch      | 0.5011  | 0.72189    | Torch     | 0.5900   | 0.45548     | Torch      | 9.908520    | 9.652524       |
| Nesterov  | 0.4023    | 0.34510     | Torch      | 0.6769  | 0.72279    | Torch     | 0.4855   | 0.45149     | Torch      | 9.602156    | 9.624092       |
| AdaGrad   | 0.3489    | 0.47579     | TF         | 0.7198  | 0.61781    | TF        | 0.4539   | 0.51806     | TF         | 9.646044    | 9.254324       |
| RMSProp   | 0.3856    | 0.36799     | Torch      | 0.6903  | 0.70441    | Torch     | 0.4913   | 0.45324     | Torch      | 9.903184    | 9.463510       |
| Adam      | 0.3992    | 0.34945     | Torch      | 0.6793  | 0.71930    | Torch     | 0.4837   | 0.46032     | Torch      | 9.348383    | 9.677068       |



| Категория                | **Keras (`tf.keras`)**                          | **PyTorch**                                     |
| ------------------------ | ----------------------------------------------- | ----------------------------------------------- |
| **Уровень абстракции**   | Высокоуровневый (ближе к "конструктору")        | Среднеуровневый (почти "ручной")                |
| **Простота кода**        | Максимально простой (`model = Sequential(...)`) | Чуть сложнее, нужно писать классы               |
| **Гибкость**             | Ограничена архитектурой Keras                   | Полная: кастомные forward, оптимизаторы, лоссы  |
| **Обучение**             | Автоматическое через `.fit()`                   | Явное через `for epoch in range(...)`           |
| **Отладка**              | Умеренная (сложно лезть внутрь `.fit()`)        | Отличная (можно пошагово дебажить `forward()`)  |
| **Визуализация**         | Встроенная: `TensorBoard`, `model.summary()`    | Через `TensorBoard`, `wandb`, `matplotlib`      |
| **Готовые модели**       | Очень много (`tf.keras.applications`)           | Много (`torchvision.models`, `torch.hub`)       |
| **Совместимость**        | Плотно связан с TensorFlow                      | Независим, но интегрируется с ONNX              |
